toolchain/parse/handle_require.cpp

```diff
@@ -0,0 +1,48 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parse/context.h"
+#include "toolchain/parse/handle.h"
+
+namespace Carbon::Parse {
+
+auto HandleRequireAfterIntroducer(Context& context) -> void {
+  auto state = context.PopState();
+  state.kind = StateKind::RequireDecl;
+  context.PushState(state);
+
+  if (context.PositionIs(Lex::TokenKind::Impls)) {
+    // impls <expression> ...
+    context.AddLeafNode(NodeKind::RequireDefaultSelfImpls, context.Consume());
+    context.PushState(StateKind::Expr);
+  } else {
+    // <expression> impls <expression>...
+    context.PushState(StateKind::RequireBeforeImpls);
+    context.PushStateForExpr(PrecedenceGroup::ForRequireImpls());
+  }
+}
+
+auto HandleRequireBeforeImpls(Context& context) -> void {
+  auto state = context.PopState();
+  if (auto impls = context.ConsumeIf(Lex::TokenKind::Impls)) {
+    context.AddNode(NodeKind::RequireTypeImpls, *impls, state.has_error);
+    context.PushState(StateKind::Expr);
+  } else {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(RequireExpectedImpls, Error,
+                        "expected `impls` in `require` declaration");
+      context.emitter().Emit(*context.position(), RequireExpectedImpls);
+    }
+    context.ReturnErrorOnState();
+  }
```

Comment:
I'm unsure and didn't see this mentioned one way or the other in the docs: should we allow any `where` requirement here (eg, `require AssocConst1.X == AssocConst2.Y;`), or only `impls` requirements?

Reply:
OK. At some point we might want to consider using the same parse nodes in both cases, but I think it's fine to not do that in this PR.

Reply:
I am having second thoughts about this.

When `Self` is in a generic parameter of an interface, it ends up in the specific for a SpecificInterface in the resulting facet type. This is nice, we have a nice well defined time and method for replacing that Self with a specific.

When it is in a requirement though then it sticks around in the facet type and in `T:! Constraint` scenario you get back to having a Self that refers to the T, and it looks like the exact same situation as for `.Self`, I think?

If only impls can be used then Self is always the type or part of a specific… IIUC